File: 21.merge-two-sorted-lists.java

import java.util.List;

/*
 * @lc app=leetcode id=21 lang=java
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode(0); // Tạo nút giả để dễ dàng thao tác
        ListNode current = dummy; // Biến current để theo dõi nút cuối cùng
        while (list1 != null && list2 != null) { // Duyệt cả hai danh sách
            if (list1.val <= list2.val) { // So sánh giá trị của hai nút
                current.next = list1; // Gắn nút từ list1 vào danh sách kết quả
                list1 = list1.next; // Di chuyển con trỏ list1
            } else {
                current.next = list2; // Gắn nút từ list2 vào danh sách kết quả
                list2 = list2.next; // Di chuyển con trỏ list2
            }
            current = current.next; // Di chuyển con trỏ current
        }
        // Nếu một trong hai danh sách còn nút, gắn phần còn lại vào danh sách kết quả
        if (list1 != null) {
            current.next = list1; // Gắn phần còn lại của list1
        } else {
            current.next = list2; // Gắn phần còn lại của list2
        }
        return dummy.next; // Trả về danh sách đã hợp nhất, bắt đầu từ nút tiếp theo của nút giả
    }
}
// @lc code=end

File: 21.merge-two-sorted-lists.py
#
# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1: # Nếu list1 rỗng
            return list2 # Trả về list2     
        if not list2: # Nếu list2 rỗng
            return list1 # Trả về list1
        if list1.val < list2.val: # Nếu giá trị của list1 nhỏ hơn list2
            list1.next = self.mergeTwoLists(list1.next, list2) # Gọi đệ quy với phần còn lại của list1 và toàn bộ list2
            return list1 # Trả về list1
        else: # Nếu giá trị của list1 lớn hơn hoặc bằng list2
            list2.next = self.mergeTwoLists(list1, list2.next) # Gọi đệ quy với toàn bộ list1 và phần còn lại của list2
            return list2 # Trả về list2
# @lc code=end

